### PR TITLE
Add current humidity to humidifier history chart

### PR DESCRIPTION
**Pull request by Shulyaka**
## Proposed change

Add current humidity to the humidifier history chart.
Also, if the `action` attribute is available, the graph is filled (shaded) depending on active/idle state instead of on/off state.
This replicates the thermostat card behavior when these new attributes are present while keeping backwards compatibility and old behavior when they are not present.

image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/resources/ha-style.ts [hunk 1]
```diff
@@ -157,6 +157,8 @@ documentContainer.innerHTML = `<custom-style>
       --state-device_tracker-home-color: var(--green-color);
       --state-fan-active-color: var(--cyan-color);
       --state-humidifier-active-color: var(--blue-color);
+      --state-humidifier-humidify-color: var(--blue-color);
+      --state-humidifier-dry-color: var(--orange-color);
```
Comment (piitaya) [inline src/resources/ha-style.ts:161]: State color doesn't apply to action, only state.
https://www.home-assistant.io/integrations/frontend/#state-color

We should rename `--state-humidifier-active-color` with `--state-humidifier-on-color` and use it for drying and humidifying.

There is a PR open for new humidifier more-info, but it's currently in review by the design team. Then, we can adjust the colors when we will have feedback on this PR for consistency.

For now, we can use the `--state-humidifier-on-color` for both actions.

File: src/components/chart/state-history-chart-line.ts [hunk 2]
```diff
@@ -328,23 +328,94 @@ class StateHistoryChartLine extends LitElement {
...
         });
       } else if (domain === "humidifier") {
+        const hasAction = states.states.some(
+          (entityState) => entityState.attributes?.action
+        );
+        const hasCurrent = states.states.some(
+          (entityState) => entityState.attributes?.current_humidity
+        );
+
+        const hasHumidifying =
```
Comment (piitaya) [inline src/components/chart/state-history-chart-line.ts:338]:
```suggestion
        const isHumidifying =
```

File: src/components/chart/state-history-chart-line.ts [hunk 3]
```diff
@@ -328,23 +328,94 @@ class StateHistoryChartLine extends LitElement {
...
+          hasAction &&
+          states.states.some(
+            (entityState: LineChartState) =>
+              entityState.attributes?.action === "humidifying"
+          );
+        const hasDrying =
```
Comment (piitaya) [inline src/components/chart/state-history-chart-line.ts:344]:
```suggestion
        const isDrying =
```

File: src/components/chart/state-history-chart-line.ts [hunk 4]
```diff
@@ -328,23 +328,94 @@ class StateHistoryChartLine extends LitElement {
...
+            true,
+            computedStyles.getPropertyValue("--state-humidifier-dry-color")
+          );
+        } else {
+          addDataSet(
+            `${this.hass.localize("ui.card.humidifier.on_entity", {
+              name: name,
+            })}`,
+            true
+          );
```
Comment (piitaya) [inline src/components/chart/state-history-chart-line.ts:392]: We must have the same logic as climate, if there is no action, we don't know if it's active or now so we must not color the area.

**Review by piitaya — CHANGES_REQUESTED** [review]

File: src/components/chart/state-history-chart-line.ts (hunk 4 above)
Comment (Shulyaka) [reply src/components/chart/state-history-chart-line.ts:392]: This is for backwards compatibility. For humidifiers that does not expose `action` attribute, we shade when it is on and don't shade when it is off. This is the current behavior, and the description (legend) will indicate what this metric is about.

File: src/resources/ha-style.ts (hunk 1 above)
Comment (Shulyaka) [reply src/resources/ha-style.ts]: Done

File: src/components/chart/state-history-chart-line.ts (hunk 2 above)
Comment (Shulyaka) [reply src/components/chart/state-history-chart-line.ts:338]: Why? I tried to keep it close to the `climate` code. `hasHumidifying` means it *has* action='humidifying' somewhere in the history. Did I get it wrong?

File: src/components/chart/state-history-chart-line.ts (hunk 4 above)
Comment (Shulyaka) [reply src/components/chart/state-history-chart-line.ts:392]: Don't get me wrong, I am OK removing it, just wanted to double check if you are sure :)

File: src/components/chart/state-history-chart-line.ts (hunk 2 above)
Comment (piitaya) [reply src/components/chart/state-history-chart-line.ts:338]: climate is using `isHeating` or `isCooling`

**Review by piitaya — APPROVED** [review]

**Comment by piitaya** [comment]
Thanks !

**Comment by Shulyaka** [comment]
Thanks for the review!

**Pull request merged**